src/crypto/shelley/transactionUtils.js

```diff
@@ -0,0 +1,98 @@
+// @flow
+import {BigNumber} from 'bignumber.js'
+import {
+  BigNum,
+  Coin,
+  LinearFee,
+  TransactionBuilder,
+} from 'react-native-haskell-shelley'
+
+import {sendAllUnsignedTx, newAdaUnsignedTx} from './transactions'
+import {NETWORKS} from '../../config/networks'
+import {CardanoError, InsufficientFunds} from '../errors'
+import {Logger} from '../../utils/logging'
+
+import type {Addressing, AddressedUtxo, BaseSignRequest} from '../types'
+
+export type CreateUnsignedTxRequest = {|
+  changeAddr: {
+    address: string,
+    ...Addressing,
+  },
+  absSlotNumber: BigNumber,
+  receiver: string,
+  addressedUtxos: Array<AddressedUtxo>,
+  ...
+    | {|
+        amount: string, // in lovelaces
+      |}
+    | {|
+        shouldSendAll: true,
+      |},
+|}
+
+export type CreateUnsignedTxResponse = BaseSignRequest<TransactionBuilder>
+
+export const createUnsignedTx = async (
+  request: CreateUnsignedTxRequest,
+): Promise<CreateUnsignedTxResponse> => {
+  Logger.debug('createUnsignedTx called', request)
+  const {changeAddr, receiver, addressedUtxos, absSlotNumber} = request
+  try {
+    const KEY_DEPOSIT = NETWORKS.HASKELL_SHELLEY.KEY_DEPOSIT
+    const LINEAR_FEE = NETWORKS.HASKELL_SHELLEY.LINEAR_FEE
+    const MINIMUM_UTXO_VAL = NETWORKS.HASKELL_SHELLEY.MINIMUM_UTXO_VAL
+    const POOL_DEPOSIT = NETWORKS.HASKELL_SHELLEY.POOL_DEPOSIT
+
+    const protocolParams = {
+      keyDeposit: await BigNum.from_str(KEY_DEPOSIT),
+      linearFee: await LinearFee.new(
+        await Coin.from_str(LINEAR_FEE.COEFFICIENT),
+        await Coin.from_str(LINEAR_FEE.CONSTANT),
+      ),
+      minimumUtxoVal: await Coin.from_str(MINIMUM_UTXO_VAL),
+      poolDeposit: await BigNum.from_str(POOL_DEPOSIT),
+    }
+
+    let unsignedTxResponse
+    if (request.shouldSendAll != null) {
+      unsignedTxResponse = await sendAllUnsignedTx(
+        receiver,
+        addressedUtxos,
+        absSlotNumber,
+        protocolParams,
+      )
+    } else if (request.amount != null) {
+      const amount = request.amount
+
+      unsignedTxResponse = await newAdaUnsignedTx(
+        [{address: receiver, amount}],
+        {
+          address: changeAddr.address,
+          addressing: changeAddr.addressing,
+        },
+        addressedUtxos,
+        absSlotNumber,
+        protocolParams,
+        [],
+      )
+    } else {
+      throw new Error('shelley::createUnsignedTx:: unknown param')
+    }
+    Logger.debug(
+      `createUnsignedTx success: ${JSON.stringify(unsignedTxResponse)}`,
+    )
+    // TODO(V-almonacid): Should we instead return a HaskellShelleyTxSignRequest
```

Comment:
I'd like to get rid of BaseSignRequest eventually and replace it with ISignRequest and different things that inherit it.

Reply:
I'm also not very happy with the way I'm handling some generic types. I've improved the code significantly to have a more flexible/generic wallet API but I feel like it can be done better.